### PR TITLE
Param values under runtime

### LINKED ISSUE #2414: explicit parameter values inside a runtime namespace
A user has complained that this:
```
[runtime]
  [[FAM<m>]]
  [[foo<m=3>]]
      inherit = FAM<m=3>
```
fails validation:
```
ERROR, parameter values not supported here: inherit = FAM<m=3>
```

The problem is, we're assuming that the content of a namespace will always pertain to the same parameter value as the heading, i.e.:
```
[runtime]
  [[FAM<m>]]
  [[foo<m=3>]]
      inherit = FAM<m>  # m is already '3' in here
```

That is probably true, but we should probably allow `inherit = FAM<m=3>` for clarity.  If necessary we could fail any `inherit = FAM<m=q>` where `q` is not 3 (in this case).

### PR DESCRIPTION
**Pull request by hjoliver**
Supersedes #2427

Close #2414

### CODE REVIEW

**Comment by matthewrmshin** [comment]
(Failing PEP8 test on Travis CI.)

File: lib/cylc/param_expand.py [hunk 1]
```diff
@@ -210,25 +210,30 @@ def _expand_name(self, str_tmpl, param_list, results, spec_vals=None):
...
-        for p_name in used_param_names:
-            msg = None
-            if '=' in p_name:
-                msg = 'values'
-            elif '-' in p_name or '+' in p_name:
-                msg = 'offsets'
-            if msg is not None:
-                raise ParamExpandError("ERROR, parameter %s not supported"
-                                       " here: %s" % (msg, origin))
+        used_params = [i.strip() for i in p_tmpl[1:-1].split(',')]
```
Comment (matthewrmshin) [inline lib/cylc/param_expand.py]: I wonder if it is worth adjusting what `REC_P_NAME` captures so we don't have to do `p_tmpl[1:-1]`. (`REC_P_NAME` is used in another location with a similar logic.)

Comment (matthewrmshin) [reply lib/cylc/param_expand.py]: I think the code will be easier to understand if we have something like `captured_params` and `used_params`.

**Comment by matthewrmshin** [comment]
(I don't understand why the failed test is failing on Travis CI. It is as if you have already merged master or #2431 into this branch.)

**Comment by matthewrmshin** [comment]
(In which case, you should have a simple block to convert the regex-captured integer string into an `int` for int-only parameters.)

**Comment by hjoliver** [comment]
(yeah that was weird - something was returning an int on Travis CI and a string in my environment.  I've rebased to master now).

Still TODO - handle and check whitespace in parameter notation.

**Comment by matthewrmshin** [comment]
@oliver-sanders please sanity check.

**Comment by oliver-sanders** [comment]
I find it a little confusing that the following two examples work and are equivalent:
```ini
[runtime]
    [[task<param=value>]]
        inherit = FAM<param>
```
```ini
[runtime]
    [[task<param=value>]]
        inherit = FAM<param=value>
```
Whereas these examples doesn't work:
```ini
[runtime]
    [[task<param>]]
        inherit = FAM<param=value>
```
```ini
[runtime]
    [[foo]]
        inherit = FAM<param=value>
```

**Comment by hjoliver** [comment]
@oliver-sanders - your first two examples make perfect sense to me, given that what gets expanded/replicated across parameter ranges is task names in the graph and corresponding task definition sections under runtime ... settings underneath a runtime namespace do not get expanded independently, they need to take the value of the parameter in the heading (either implicitly - as before the PR - or explicitly - as on the PR). 

I'm trying to figure out if your 3rd and 4th examples makes sense: (3) is it plausible to have all instances of a parameterized task inherit from a single instance of a family that is expanded over the same parameter?  and (4) is there a case for a single unparameterized task that inherits from a specific instance of a parameterized family?  Maybe we should support those in principle even if they seem unlikely in practice.  (4) suggests this though, which is definitely nonsensical:
```ini
[runtime]
    [[foo]]
          inherit = FAM<param>
```

**Comment by oliver-sanders** [comment]
> I'm trying to figure out if your 3rd and 4th examples makes sense

Probably not, though once we set the president that `<param=value>` can be used anywhere I wouldn't be surprised if we started to see it used everywhere.

> is there a case for a single unparameterized task that inherits from a specific instance of a parameterized family

Maybe?

```ini
[cylc]
    [[parameters]]
        transport = car, train, plane
[scheduling]
    [[dependencies]]
        graph = """
            start => travel<transport>
            airport_security => travel<transport=plane>
        """
[runtime]
    [start]

    [TRANSPORT<transport>]
        script = # mv
    [TRANSPORT<transport=plane>]
        [[[environment]]]
            FLIGHT_NUMBER=abc001
            TERMINAL=3

    [travel<transport>]
        inherit = travel<transport>
    [airport_security]
        inherit = TRANSPORT<transport=plane>
```

**Comment by oliver-sanders** [comment]
Coming back to this we could just put in an error message to cover the latter two cases. The stock error is somewhat misleading:

```
ERROR, parameter values not supported here: inherit = FAM<param=1>
```
```
ERROR, undefined parent for foo: FAM<param=1>
```

**Comment by hjoliver** [comment]
@oliver-sanders - based on your example (which is mildly convincing!) I've decided to allow inheritance from any specific parameterized parent.  The value of parameters in the parent can either be explicit (`parent<m=3>`) in which case any legal value of the parameter is OK, regardless of whether the namespace heading is expanded over that parameter or not; or implicit (`parent<m>`) in which case the value must be supplied by expansion of the same parameter in the namespace heading ... but anything else is illegal, with better error messages.  Tests updated accordingly.

**Pull request merged**